central/baseimage/datastore/datastore_impl.go

```diff
@@ -0,0 +1,155 @@
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+type datastoreImpl struct {
+	storage baseImageStore.Store
+	db      postgres.DB
+}
+
+var (
+	log       = logging.LoggerForModule()
+	imagesSAC = sac.ForResource(resources.ImageAdministration)
+)
+
+const (
+	baseImagesTable = "base_images"
+	// The 'firstlayerdigest' column is indexed.
+	listByFirstLayerQuery = "SELECT id FROM " + baseImagesTable + " WHERE firstlayerdigest = $1"
+
+	// Query to find the ID by manifest digest
+	getByManifestDigestQuery = "SELECT id FROM " + baseImagesTable + " WHERE manifestdigest = $1"
+)
+
+// New creates a new DataStore instance backed by PostgreSQL.
+func New(store baseImageStore.Store, db postgres.DB) DataStore {
+	return &datastoreImpl{
+		storage: store,
+		db:      db,
+	}
+}
+
+func (ds *datastoreImpl) UpsertImage(ctx context.Context, image *storage.BaseImage, layers []*storage.BaseImageLayer) error {
+	if ok, err := imagesSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	for i, layer := range layers {
+		if layer.GetId() == "" {
+			layer.Id = uuid.NewV4().String()
+		}
+
+		layer.BaseImageId = image.GetId()
+		layer.Index = int32(i)
+	}
+	if len(layers) > 0 {
+		log.Warnf("FirstLayerDigest mismatch for image %s: claims %s but first layer is %s.",
+			image.GetId(), image.GetFirstLayerDigest(), layers[0].GetLayerDigest())
+		image.FirstLayerDigest = layers[0].GetLayerDigest()
+	}
+
+	image.Layers = layers
+	return ds.storage.Upsert(ctx, image)
+}
+
+func (ds *datastoreImpl) UpsertImages(ctx context.Context, imagesWithLayers map[*storage.BaseImage][]*storage.BaseImageLayer) error {
+	if ok, err := imagesSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+	batch := make([]*storage.BaseImage, 0, len(imagesWithLayers))
+
+	for image, layers := range imagesWithLayers {
+		// Attach layers to the parent
+		image.Layers = layers
+
+		if len(layers) > 0 && image.GetFirstLayerDigest() != layers[0].GetLayerDigest() {
+			log.Warnf("FirstLayerDigest mismatch for image %s: claims %s but first layer is %s.",
+				image.GetId(), image.GetFirstLayerDigest(), layers[0].GetLayerDigest())
+			image.FirstLayerDigest = layers[0].GetLayerDigest()
+		}
+
+		for i, layer := range image.GetLayers() {
+			if layer.GetId() == "" {
+				layer.Id = uuid.NewV4().String()
+			}
+			layer.BaseImageId = image.GetId()
+			layer.Index = int32(i)
+		}
+
+		batch = append(batch, image)
+	}
+
+	return ds.storage.UpsertMany(ctx, batch)
+}
+
+func (ds *datastoreImpl) ListCandidateBaseImages(ctx context.Context, firstLayer string) ([]*storage.BaseImage, error) {
+	if ok, err := imagesSAC.ReadAllowed(ctx); err != nil {
+		return nil, err
+	} else if !ok {
+		return nil, sac.ErrResourceAccessDenied
+	}
+	rows, err := ds.db.Query(ctx, listByFirstLayerQuery, firstLayer)
```

Comment:
Why don't we retrieve the whole base image row in this query, and we can avoid the `GetMany` call later?

Reply:
I investigated a bit. `GetMany` ensures to get base image object defined in proto, including base image layers. I mean it should be a safer way to retrieve things.